src/data_readers/data_reader_hdf5.cpp

```diff
@@ -86,13 +124,73 @@ namespace lbann {
 
     //todo add error checking
     H5Sselect_hyperslab(filespace, H5S_SELECT_SET, offset, NULL, count, dims_local);
-    H5Dread(h_data, H5T_NATIVE_SHORT, memspace, filespace, H5P_DEFAULT, data_out);
+    sample.set(conduit::DataType::uint16(xPerNode * yPerNode * zPerNode * cPerNode));
+
+    unsigned short* buf = sample.value();
+    H5Dread(h_data, H5T_NATIVE_SHORT, memspace, filespace, H5P_DEFAULT, buf);
+  }
+
+  void hdf5_reader::read_hdf5_sample(int data_id, conduit::Node& sample) {
+    //int world_rank = get_rank_in_world();
+    int world_rank = dc::get_input_rank(*get_comm()); // Should probably be trainer rank
+    auto file = m_file_paths[data_id];
+    hid_t h_file = H5Fopen(file.c_str(), H5F_ACC_RDONLY, m_fapl);
+
+#if 0
+    dc::MPIPrintStreamInfo() << "HDF5 file opened: "
+                             << file;
+#endif
+
+    if (h_file < 0) {
+      throw lbann_exception(std::string{} + __FILE__ + " " + std::to_string(__LINE__) +
+                            " hdf5_reader::load() - can't open file : " + file);
+    }
+
+    // load in dataset
+    hid_t h_data = CHECK_HDF5(
+                              H5Dopen(h_file, HDF5_KEY_DATA.c_str(), H5P_DEFAULT));
+    hid_t filespace = CHECK_HDF5(H5Dget_space(h_data));
+    //get the number of dimesnionse from the dataset
+    int rank1 = H5Sget_simple_extent_ndims(filespace);
+    hsize_t dims[rank1];
+    // read in what the dimensions are
+    CHECK_HDF5(H5Sget_simple_extent_dims(filespace, dims, NULL));
+
+    if (h_data < 0) {
+      throw lbann_exception(std::string{} + __FILE__ + " " + std::to_string(__LINE__) +
+                            " hdf5_reader::load() - can't find hdf5 key : " + HDF5_KEY_DATA);
+    }
+
+    const std::string conduit_obj = LBANN_DATA_ID_STR(data_id);
+    conduit::Node slab;
+    read_hdf5_hyperslab(h_data, filespace, world_rank, HDF5_KEY_DATA, dims, slab);
+    sample[conduit_obj+"/slab"] = slab;
+    //close data set
+    H5Dclose(h_data);
+
+    if (m_has_responses) {
+      h_data = H5Dopen(h_file, HDF5_KEY_RESPONSES.c_str(), H5P_DEFAULT);
+      H5Dread(h_data, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL, H5P_DEFAULT, m_all_responses);
+      // conduit::Node work;
+      // conduit::relay::io::hdf5_read(h_data, HDF5_KEY_RESPONSES.c_str(), work);
+      //      node[conduit_obj+"/responses"].set(conduit::DataType::c_float(4));
+      sample[conduit_obj+"/responses"].set(m_all_responses, 4);
+      H5Dclose(h_data);
+    }
+    H5Fclose(h_file);
+    return;
   }
 
   void hdf5_reader::load() {
     lbann_comm* l_comm = get_comm();
+    if(dc::get_rank_stride() != 1) {
+      LBANN_ERROR("HDF5 MPI-IO data reader requires DistConv rank stride = 1");
```

Comment:
Is there any reason with this requirement? I've been testing it with `rank_stride` == `num_partitions`.